server/routes/login.js

```diff
@@ -210,16 +36,86 @@ module.exports = () => (router) => {
 
       res.render('pages/index', data)
     }
-  })
+  }
+
+  router.get('/', postLogin)
+
+  router.get('/auth/login', postLogin)
 
+  router.post('/auth/2fa', (req, res) => {
+    if (validateMFA(req.body.code, req.user.username) === true) {
+      processLogin(req)
+      res.redirect(`/calendar/${utilities.getStartMonth()}`)
+    } else {
+      logError(req.url, '2FA failure')
+      res.render('pages/two-factor-auth', { authError: true, csrfToken: res.locals.csrfToken })
+    }
+  })
   return router
 }
 
-async function getStaffMemberEmployeeName(apiUrl, startMonth, accessToken) {
-  const staffMemberResponse = await staffMemberService.getStaffMemberData(apiUrl, startMonth, accessToken)
+function processLogin(req) {
+  req.user.employeeName = jwtDecode(req.user.token).name
+  userAuthenticationService.updateUserSessionExpiryAndLastLoginDateTime(
+    req.user.username,
+    new Date(Date.now() + config.hmppsCookie.expiryMinutes * 60 * 1000),
+  )
+}
+
+function validateMFA(code, username) {
+  const userDetails = userAuthenticationService.getUserAuthenticationDetails(username)[0]
+  return utilities.createTwoFactorAuthenticationHash(code) === userDetails.TwoFactorAuthenticationHash
+}
+
+function userHasRole(token, role) {
+  jwtDecode(token).authorities.includes(role)
+}
+
+function isQuantumAddress(req) {
+  return ipRangeCheck(getIpAddress(req), config.quantumAddresses.split(','))
+}
+
+function getIpAddress(req) {
+  return (
+    req.headers['x-forwarded-for'] ||
+    req.connection.remoteAddress ||
+    req.socket.remoteAddress ||
+    (req.connection.socket ? req.connection.socket.remoteAddress : null)
+  )
+}
+
+function sendMFA(userName, userAuthentication) {
+  const twofactorCode = utilities.get2faCode()
+
+  userAuthenticationService.updateTwoFactorAuthenticationHash(
+    userName,
+    utilities.createTwoFactorAuthenticationHash(twofactorCode.toString()),
+  )
 
-  if (staffMemberResponse !== null) {
-    return staffMemberResponse.staffMembers[0].employeeName
+  const {
+    notify: { url, clientKey, smsTemplateId, emailTemplateId },
+  } = config
+  const notify = url ? new NotifyClient(url, clientKey) : new NotifyClient(clientKey)
+
+  if (userAuthentication.UseSms) {
+    // For SMS
+    notify
+      .sendSms(smsTemplateId, userAuthentication.Sms || '', {
+        personalisation: { '2fa_code': twofactorCode },
+      })
+      .catch((err) => {
+        throw new Error(err)
```

Comment:
Why are you throwing an error here? Is this to be caught by error handling middleware?